README.md
# Kowala Currency Plugin for Edge
[![js-standard-style](https://cdn.rawgit.com/feross/standard/master/badge.svg)](https://github.com/feross/standard)

Implements Kowala send/receive functionality per the spec for crypto currency plugins for [airbitz-core-js](https://github.com/Airbitz/airbitz-core-js)

## Installing

    npm i git://github.com/kowala-tech/edge-currency-kusd.git#develop -s

```
import { kusdCurrencyPluginFactory } from `edge-currency-kusd`
```

Now you can pass `kusdCurrencyPluginFactory` to `edge-core-js`.

```
const context = makeEdgeContext({
  apiKey: YOUR_API_KEY,
  plugins: [ kusdCurrencyPluginFactory ]
})
```

## Contributing

You'll need to install Yarn 1.3.2 globally on your machine

To run a local version of this repo inside the full Edge Wallet app, clone this repo at the same level as `edge-react-gui`

    git clone git@github.com:Airbitz/edge-currency-kusd.git`
    cd edge-currency-kusd
    yarn

Run `npm run test` to run the unit tests.

To use the local cloned version of this repo, `cd edge-react-gui` and run

    npm run updot edge-currency-kusd
    npm run postinstall

This will copy the necessary files from `edge-currency-kusd` into the `edge-react-gui/node_modules/edge-currency-kusd` replacing the npm installed version. This needs to be done after any modifications to `edge-currency-kusd`

## License
BSD 3


CHANGELOG.md
# edge-currency-kusd

## 0.10.1

* Support the `token_info` prefix inside token URI's.

## 0.10.0

* Fix progress ratio reporting.
* Add Hurify & Indorse tokens by default.

## 0.9.0

* Parse the Add Token URI scheme

## 0.8.0

* Change gasPrice to be in Gwei

## 0.7.12

* Broadcast transactions to both Etherscan and Blockcypher for reliability.

## 0.4.0

* Upgraded to the latest airbitz-core-types, which add a `pluginName` to the factory.


test/engine/engine.js
// @flow
import EventEmitter from 'events'

import { makeFakeIos, makeContext, destroyAllContexts } from 'edge-core-js'
import type {
  // EdgeSpendInfo,
  EdgeWalletInfo,
  EdgeCurrencyEngineOptions,
  EdgeCurrencyEngineCallbacks
} from 'edge-core-js'
import { assert } from 'chai'
import { describe, it } from 'mocha'
import fetch from 'node-fetch'
import request from 'request'

import * as Factories from '../../src/index.js'
import fixtures from './fixtures.json'

// const DATA_STORE_FOLDER = 'txEngineFolderBTC'

for (const fixture of fixtures) {
  const CurrencyPluginFactory = Factories[fixture['factory']]
  const WALLET_TYPE = fixture['WALLET_TYPE']
  // const TX_AMOUNT = fixture['TX_AMOUNT']

  let plugin, keys, engine
  const emitter = new EventEmitter()
  const [fakeIo] = makeFakeIos(1)
  if (!fakeIo.folder) {
    throw new Error('Missing fakeio.folder')
  }
  const walletLocalFolder = fakeIo.folder
  const plugins = [CurrencyPluginFactory]
  // $FlowFixMe
  fakeIo.fetch = fetch

  const context = makeContext({ io: fakeIo, plugins })

  const callbacks: EdgeCurrencyEngineCallbacks = {
    onAddressesChecked (progressRatio) {
      // console.log('onAddressesCheck', progressRatio)
      emitter.emit('onAddressesCheck', progressRatio)
    },
    onTxidsChanged (txid) {
      // console.log('onTxidsChanged', txid)
      emitter.emit('onTxidsChanged', txid)
    },
    onBalanceChanged (currencyCode, balance) {
      // console.log('onBalanceChange:', currencyCode, balance)
      emitter.emit('onBalanceChange', currencyCode, balance)
    },
    onBlockHeightChanged (height) {
      // console.log('onBlockHeightChange:', height)
      emitter.emit('onBlockHeightChange', height)
    },
    onTransactionsChanged (transactionList) {
      // console.log('onTransactionsChanged:', transactionList)
      emitter.emit('onTransactionsChanged', transactionList)
    }
  }

  const currencyEngineOptions: EdgeCurrencyEngineOptions = {
    callbacks,
    walletLocalFolder,
    walletLocalEncryptedFolder: walletLocalFolder
  }

  describe(`Create Plugin for Wallet type ${WALLET_TYPE}`, function () {
    it('Plugin', function () {
      return context.getCurrencyPlugins().then(currencyPlugins => {
        const currencyPlugin = currencyPlugins[0]
        assert.equal(
          currencyPlugin.currencyInfo.currencyCode,
          fixture['Test Currency code']
        )
        plugin = currencyPlugin
        keys = plugin.createPrivateKey(WALLET_TYPE)
        const info: EdgeWalletInfo = {
          id: '1',
          type: WALLET_TYPE,
          keys
        }
        keys = plugin.derivePublicKey(info)
      })
    })
  })

  describe(`Make Engine for Wallet type ${WALLET_TYPE}`, function () {
    // before('Create local cache file', function (done) {
    //   walletLocalFolder
    //     .folder(DATA_STORE_FOLDER)
    //     .file('addresses.json')
    //     .setText(JSON.stringify(dummyAddressData))
    //     .then(() =>
    //       walletLocalFolder
    //         .folder(DATA_STORE_FOLDER)
    //         .file('txs.json')
    //         .setText(JSON.stringify(dummyTransactionsData))
    //     )
    //     .then(() =>
    //       walletLocalFolder
    //         .folder(DATA_STORE_FOLDER)
    //         .file('headers.json')
    //         .setText(JSON.stringify(dummyHeadersData))
    //     )
    //     .then(done)
    // })

    it('Make Engine', function () {
      const info: EdgeWalletInfo = {
        id: '1',
        type: WALLET_TYPE,
        keys
      }
      return plugin
        .makeEngine(info, currencyEngineOptions)
        .then(e => {
          engine = e
          assert.equal(typeof engine.startEngine, 'function', 'startEngine')
          assert.equal(typeof engine.killEngine, 'function', 'killEngine')
          // assert.equal(typeof engine.enableTokens, 'function', 'enableTokens')
          assert.equal(
            typeof engine.getBlockHeight,
            'function',
            'getBlockHeight'
          )
          assert.equal(typeof engine.getBalance, 'function', 'getBalance')
          assert.equal(
            typeof engine.getNumTransactions,
            'function',
            'getNumTransactions'
          )
          assert.equal(
            typeof engine.getTransactions,
            'function',
            'getTransactions'
          )
          assert.equal(
            typeof engine.getFreshAddress,
            'function',
            'getFreshAddress'
          )
          assert.equal(
            typeof engine.addGapLimitAddresses,
            'function',
            'addGapLimitAddresses'
          )
          assert.equal(typeof engine.isAddressUsed, 'function', 'isAddressUsed')
          assert.equal(typeof engine.makeSpend, 'function', 'makeSpend')
          assert.equal(typeof engine.signTx, 'function', 'signTx')
          assert.equal(typeof engine.broadcastTx, 'function', 'broadcastTx')
          assert.equal(typeof engine.saveTx, 'function', 'saveTx')
          return true
        })
    })
  })

  // describe(`Is Address Used for Wallet type ${
  //   WALLET_TYPE
  // } from cache`, function () {
  //   it('Checking a wrong formated address', function (done) {
  //     try {
  //       engine.isAddressUsed('TestErrorWithWrongAddress')
  //     } catch (e) {
  //       assert(e, 'Should throw')
  //       assert.equal(e.message, 'Wrong formatted address')
  //       done()
  //     }
  //   })

  //   it("Checking an address we don't own", function () {
  //     try {
  //       assert.equal(
  //         engine.isAddressUsed('mnSmvy2q4dFNKQF18EBsrZrS7WEy6CieEE'),
  //         false
  //       )
  //     } catch (e) {
  //       assert(e, 'Should throw')
  //       assert.equal(e.message, 'Address not found in wallet')
  //     }
  //   })

  //   it('Checking an empty P2SH address', function (done) {
  //     assert.equal(
  //       engine.isAddressUsed('2N9DbpGaQEeLLZgPQP4gc9oKkrFHdsj5Eew'),
  //       false
  //     )
  //     done()
  //   })

  //   it('Checking a non empty P2SH address 1', function (done) {
  //     assert.equal(
  //       engine.isAddressUsed('2MwLo2ghJeXTgpDccHGcsTbdS9YVfM3K5GG'),
  //       true
  //     )
  //     done()
  //   })

  //   it('Checking a non empty P2SH address 2', function (done) {
  //     assert.equal(
  //       engine.isAddressUsed('2MxRjw65NxR4DsRj2z1f5xFnKkU5uMRCsoT'),
  //       true
  //     )
  //     done()
  //   })

  //   it('Checking a non empty P2SH address 3', function (done) {
  //     assert.equal(
  //       engine.isAddressUsed('2MxvxJh44wq17vhzGqFcAsuYsVmdEJKWuFV'),
  //       true
  //     )
  //     done()
  //   })
  // })

  // describe(`Get Transactions from Wallet type ${WALLET_TYPE}`, function () {
  //   it('Should get number of transactions from cache', function (done) {
  //     assert.equal(
  //       engine.getNumTransactions(),
  //       TX_AMOUNT,
  //       `should have ${TX_AMOUNT} tx from cache`
  //     )
  //     done()
  //   })

  //   it('Should get transactions from cache', function (done) {
  //     engine.getTransactions().then(txs => {
  //       assert.equal(
  //         txs.length,
  //         TX_AMOUNT,
  //         `should have ${TX_AMOUNT} tx from cache`
  //       )
  //       done()
  //     })
  //   })

  //   it('Should get transactions from cache with options', function (done) {
  //     engine.getTransactions({ startIndex: 1, numEntries: 2 }).then(txs => {
  //       assert.equal(txs.length, 2, 'should have 2 tx from cache')
  //       done()
  //     })
  //   })
  // })

  describe('Start engine', function () {
    it('Get BlockHeight', function (done) {
      this.timeout(10000)
      request.get(
        'http://ec2-54-159-86-65.compute-1.amazonaws.com:8080/api/blockheight',
        (err, res, body) => {
          assert(!err, 'getting block height from a second source')
          emitter.once('onBlockHeightChange', height => {
            const thirdPartyHeight = parseInt(JSON.parse(body).block_height, 8)
            assert(height >= thirdPartyHeight, 'Block height')
            assert(engine.getBlockHeight() >= thirdPartyHeight, 'Block height')
            done() // Can be "done" since the promise resolves before the event fires but just be on the safe side
          })
          engine.startEngine().catch(e => {
            console.log('startEngine error', e, e.message)
          })
        }
      )
    })
  })

  // describe(`Get Fresh Address for Wallet type ${WALLET_TYPE}`, function () {
  //   it('Should provide a non used BTC address when no options are provided', function (
  //     done
  //   ) {
  //     setTimeout(() => {
  //       const address = engine.getFreshAddress()
  //       request.get(
  //         `https://api.blocktrail.com/v1/tBTC/address/${
  //           address.publicAddress
  //         }?api_key=MY_APIKEY`,
  //         (err, res, body) => {
  //           const thirdPartyBalance = parseInt(JSON.parse(body).received)
  //           assert(!err, 'getting address incoming txs from a second source')
  //           assert(thirdPartyBalance === 0, 'Should have never received coins')
  //           done()
  //         }
  //       )
  //     }, 1000)
  //   })
  // })

  // describe(`Make Spend and Sign for Wallet type ${WALLET_TYPE}`, function () {
  //   it('Should build transaction with low fee', function () {
  //     // $FlowFixMe
  //     const templateSpend: EdgeSpendInfo = {
  //       networkFeeOption: 'low',
  //       metadata: {
  //         name: 'Transfer to College Fund',
  //         category: 'Transfer:Wallet:College Fund'
  //       },
  //       spendTargets: [
  //         {
  //           currencyCode: 'BTC',
  //           publicAddress: '2MutAAY6tW2HEyrhSadT1aQhP4KdCAKkC74',
  //           nativeAmount: '210000' // 0.021 BTC
  //         },
  //         {
  //           currencyCode: 'BTC',
  //           publicAddress: 'tb1qzu5e2xhmh7lyfs38yq0u7xmem37ufp6tp6uh6q',
  //           nativeAmount: '420000' // 0.042 BTC
  //         }
  //       ]
  //     }
  //     // $FlowFixMe
  //     return engine
  //       .makeSpend(templateSpend)
  //       .then(a => {
  //         return engine.signTx(a)
  //       })
  //       .then(a => {
  //         // console.log('sign', a)
  //       })
  //   })

  //   it('Should build transaction with low standard fee', function () {
  //     // $FlowFixMe
  //     const templateSpend: EdgeSpendInfo = {
  //       networkFeeOption: 'standard',
  //       metadata: {
  //         name: 'Transfer to College Fund',
  //         category: 'Transfer:Wallet:College Fund'
  //       },
  //       spendTargets: [
  //         {
  //           currencyCode: 'BTC',
  //           publicAddress: '2MutAAY6tW2HEyrhSadT1aQhP4KdCAKkC74',
  //           nativeAmount: '17320'
  //         }
  //       ]
  //     }
  //     return engine
  //       .makeSpend(templateSpend)
  //       .then(a => {
  //         return engine.signTx(a)
  //       })
  //       .then(a => {
  //         // console.log('sign', a)
  //       })
  //   })

  //   it('Should build transaction with middle standard fee', function () {
  //     // $FlowFixMe
  //     const templateSpend: EdgeSpendInfo = {
  //       networkFeeOption: 'standard',
  //       metadata: {
  //         name: 'Transfer to College Fund',
  //         category: 'Transfer:Wallet:College Fund'
  //       },
  //       spendTargets: [
  //         {
  //           currencyCode: 'BTC',
  //           publicAddress: '2MutAAY6tW2HEyrhSadT1aQhP4KdCAKkC74',
  //           nativeAmount: '43350000'
  //         }
  //       ]
  //     }
  //     return engine
  //       .makeSpend(templateSpend)
  //       .then(a => {
  //         return engine.signTx(a)
  //       })
  //       .then(a => {
  //         // console.log('sign', a)
  //       })
  //   })

  //   it('Should build transaction with high standard fee', function () {
  //     // $FlowFixMe
  //     const templateSpend: EdgeSpendInfo = {
  //       networkFeeOption: 'standard',
  //       metadata: {
  //         name: 'Transfer to College Fund',
  //         category: 'Transfer:Wallet:College Fund'
  //       },
  //       spendTargets: [
  //         {
  //           currencyCode: 'BTC',
  //           publicAddress: '2MutAAY6tW2HEyrhSadT1aQhP4KdCAKkC74',
  //           nativeAmount: '86700000'
  //         },
  //         {
  //           currencyCode: 'BTC',
  //           publicAddress: 'tb1qzu5e2xhmh7lyfs38yq0u7xmem37ufp6tp6uh6q',
  //           nativeAmount: '420000' // 0.042 BTC
  //         }
  //       ]
  //     }
  //     return engine
  //       .makeSpend(templateSpend)
  //       .then(a => {
  //         return engine.signTx(a)
  //       })
  //       .then(a => {
  //         // console.log('sign', a)
  //       })
  //   })

  //   it('Should build transaction with high fee', function () {
  //     // $FlowFixMe
  //     const templateSpend: EdgeSpendInfo = {
  //       networkFeeOption: 'high',
  //       metadata: {
  //         name: 'Transfer to College Fund',
  //         category: 'Transfer:Wallet:College Fund'
  //       },
  //       spendTargets: [
  //         {
  //           currencyCode: 'BTC',
  //           publicAddress: '2MutAAY6tW2HEyrhSadT1aQhP4KdCAKkC74',
  //           nativeAmount: '210000' // 0.021 BTC
  //         },
  //         {
  //           currencyCode: 'BTC',
  //           publicAddress: 'tb1qzu5e2xhmh7lyfs38yq0u7xmem37ufp6tp6uh6q',
  //           nativeAmount: '420000' // 0.042 BTC
  //         }
  //       ]
  //     }
  //     return engine
  //       .makeSpend(templateSpend)
  //       .then(a => {
  //         return engine.signTx(a)
  //       })
  //       .then(a => {
  //         // console.log('sign', a)
  //       })
  //   })

  //   it('Should build transaction with custom fee', function () {
  //     const templateSpend: EdgeSpendInfo = {
  //       networkFeeOption: 'custom',
  //       customNetworkFee: '1000',
  //       metadata: {
  //         name: 'Transfer to College Fund',
  //         category: 'Transfer:Wallet:College Fund'
  //       },
  //       spendTargets: [
  //         {
  //           currencyCode: 'BTC',
  //           publicAddress: '2MutAAY6tW2HEyrhSadT1aQhP4KdCAKkC74',
  //           nativeAmount: '210000' // 0.021 BTC
  //         },
  //         {
  //           currencyCode: 'BTC',
  //           publicAddress: 'tb1qzu5e2xhmh7lyfs38yq0u7xmem37ufp6tp6uh6q',
  //           nativeAmount: '420000' // 0.042 BTC
  //         }
  //       ]
  //     }
  //     // $FlowFixMe
  //     return engine
  //       .makeSpend(templateSpend)
  //       .then(function (a) {
  //         // console.log('makeSpend', a)
  //         return engine.signTx(a)
  //       })
  //       .then(function (a) {
  //         // console.log('signTx', a)
  //         // console.log('signTx', a.otherParams.bcoinTx.inputs)
  //       })
  //   })

  //   it('Should throw InsufficientFundsError', function () {
  //     // $FlowFixMe
  //     const templateSpend: EdgeSpendInfo = {
  //       networkFeeOption: 'high',
  //       metadata: {
  //         name: 'Transfer to College Fund',
  //         category: 'Transfer:Wallet:College Fund'
  //       },
  //       spendTargets: [
  //         {
  //           currencyCode: 'BTC',
  //           publicAddress: '2MutAAY6tW2HEyrhSadT1aQhP4KdCAKkC74',
  //           nativeAmount: '2100000000' // 0.021 BTC
  //         },
  //         {
  //           currencyCode: 'BTC',
  //           publicAddress: 'tb1qzu5e2xhmh7lyfs38yq0u7xmem37ufp6tp6uh6q',
  //           nativeAmount: '420000' // 0.042 BTC
  //         }
  //       ]
  //     }
  //     // $FlowFixMe
  //     return engine
  //       .makeSpend(templateSpend)
  //       .catch(e => assert.equal(e.message, 'InsufficientFundsError'))
  //   })

  //   after('Stop the engine', function (done) {
  //     console.log('kill engine')
  //     engine.killEngine().then(done)
  //   })
  // })
  describe('Stop the engine', function () {
    it('Should stop the engine', function (done) {
      engine.killEngine().then(() => {
        destroyAllContexts()
        done()
      })
    })
  })
}


src/types.js
/**
 * Created by paul on 8/26/17.
 */
// @flow

import type { EdgeTransaction } from 'edge-core-js'
import { currencyInfo } from './currencyInfo.js'
export const DATA_STORE_FOLDER = 'txEngineFolder'
export const DATA_STORE_FILE = 'walletLocalData.json'
export const PRIMARY_CURRENCY = currencyInfo.currencyCode

export type Settings = {
  apiServers: Array<string>,
  iosAllowedTokens: {[currencyCode: string]: boolean}
}

type FeesGasLimit = {
  regularTransaction: string,
  tokenTransaction: string
}

export type FeesGasPrice = {
  lowFee: string,
  standardFeeLow: string,
  standardFeeHigh: string,

  // The amount of wei which will be charged the standardFeeLow
  standardFeeLowAmount: string,

  // The amount of wei which will be charged the standardFeeHigh
  standardFeeHighAmount: string,
  highFee: string
}

export type Fee = {
  gasLimit: FeesGasLimit, gasPrice?: FeesGasPrice
}

export type Fees = {
  [address: string]: Fee
}

export type CalcedFees = {
  gasPrice: string,
  gasLimit: string
}

const defaultNetworkFees = {
  default: {
    gasLimit: {
      regularTransaction: '21001',
      tokenTransaction: '37123'
    },
    gasPrice: {
      lowFee: '1000000001',
      standardFeeLow: '40000000001',
      standardFeeHigh: '300000000001',
      standardFeeLowAmount: '100000000000000000',
      standardFeeHighAmount: '10000000000000000000',
      highFee: '40000000001'
    }
  },
  '1983987abc9837fbabc0982347ad828': {
    gasLimit: {
      regularTransaction: '21002',
      tokenTransaction: '37124'
    },
    gasPrice: {
      lowFee: '1000000002',
      standardFeeLow: '40000000002',
      standardFeeHigh: '300000000002',
      standardFeeLowAmount: '200000000000000000',
      standardFeeHighAmount: '20000000000000000000',
      highFee: '40000000002'
    }
  },
  '2983987abc9837fbabc0982347ad828': {
    gasLimit: {
      regularTransaction: '21002',
      tokenTransaction: '37124'
    }
  }
}

export type CustomToken = {
  currencyCode: string,
  currencyName: string,
  multiplier: string,
  contractAddress: string
}

export class WalletLocalData {
  blockHeight: number
  lastAddressQueryHeight: number
  nextNonce: string
  kusdtestnetAddress: string
  totalBalances: {[currencyCode: string]: string}
  enabledTokens: Array<string>
  transactionsObj: {[currencyCode: string]: Array<EdgeTransaction>}
  networkFees: Fees

  constructor (jsonString: string | null) {
    this.blockHeight = 0

    const totalBalances:{[currencyCode: string]: string} = {}
    this.totalBalances = totalBalances

    this.nextNonce = '0'

    this.lastAddressQueryHeight = 0

    // Dumb extra local var needed to make Flow happy
    const transactionsObj:{[currencyCode: string]: Array<EdgeTransaction>} = {}
    this.transactionsObj = transactionsObj

    this.networkFees = defaultNetworkFees

    this.kusdtestnetAddress = ''
    this.enabledTokens = [ PRIMARY_CURRENCY ]
    if (jsonString !== null) {
      const data = JSON.parse(jsonString)

      if (typeof data.blockHeight === 'number') this.blockHeight = data.blockHeight
      if (typeof data.lastAddressQueryHeight === 'string') this.lastAddressQueryHeight = data.lastAddressQueryHeight
      if (typeof data.nextNonce === 'string') this.nextNonce = data.nextNonce
      if (typeof data.kusdtestnetAddress === 'string') this.kusdtestnetAddress = data.kusdtestnetAddress
      if (typeof data.totalBalances !== 'undefined') this.totalBalances = data.totalBalances
      if (typeof data.enabledTokens !== 'undefined') this.enabledTokens = data.enabledTokens
      if (typeof data.networkFees !== 'undefined') this.networkFees = data.networkFees
      if (typeof data.transactionsObj !== 'undefined') this.transactionsObj = data.transactionsObj
    }
  }
}


src/currencyInfo.js
/* global */
// @flow

import type { EdgeCurrencyInfo } from 'edge-core-js'
import type { Settings } from './types'

const otherSettings:Settings = {
  apiServers: [
    'http://ec2-54-159-86-65.compute-1.amazonaws.com:8080'
  ],
  iosAllowedTokens: {}
}

const defaultSettings:any = {
  customFeeSettings: ['gasLimit', 'gasPrice'],
  otherSettings
}

export const currencyInfo: EdgeCurrencyInfo = {
  // Basic currency information:
  currencyCode: 'KUSD',
  currencyName: 'kUSD',
  pluginName: 'kusdtestnet',
  walletTypes: [
    'wallet:kusdtestnet'
  ],

  defaultSettings,

  addressExplorer: '',
  transactionExplorer: '',

  denominations: [
    // An array of Objects of the possible denominations for this currency
    {
      name: 'KUSD',
      multiplier: '1000000000000000000',
      symbol: 'k'
    }
  ],
  symbolImage: '', // Base64 encoded png image of the currency symbol (optional)
  symbolImageDarkMono: '', // Base64 encoded png image of the currency symbol (optional)
  metaTokens: []
}


src/index.js
/**
 * Created by paul on 8/8/17.
 */
// @flow
import { currencyInfo } from './currencyInfo.js'
import { Engine } from './currencyEngine.js'
import { DATA_STORE_FILE, DATA_STORE_FOLDER, WalletLocalData } from './types.js'
import type {
  EdgeCurrencyEngine,
  EdgeCurrencyEngineOptions,
  EdgeParsedUri,
  EdgeEncodeUri,
  EdgeCurrencyPlugin,
  EdgeCurrencyPluginFactory,
  EdgeWalletInfo
} from 'edge-core-js'
import { parse, serialize } from 'uri-js'
import { bns } from 'biggystring'
import { BN } from 'bn.js'
// import { CurrencyInfoScheme } from './schema.js'

export { calcMiningFee } from './miningFees.js'

const Buffer = require('buffer/').Buffer
const walletUtils = require('../lib/export-fixes-bundle.js').Wallet
const Util = require('../lib/export-fixes-bundle.js').Util

let io

const randomBuffer = (size) => {
  const array = io.random(size)
  return Buffer.from(array)
}

function getDenomInfo (denom: string) {
  return currencyInfo.denominations.find(element => {
    return element.name === denom
  })
}

function hexToBuf (hex: string) {
  const noHexPrefix = hex.replace('0x', '')
  const noHexPrefixBN = new BN(noHexPrefix, 16)
  const array = noHexPrefixBN.toArray()
  const buf = Buffer.from(array)
  return buf
}

function getParameterByName (param, url) {
  const name = param.replace(/[[\]]/g, '\\$&')
  const regex = new RegExp('[?&]' + name + '(=([^&#]*)|&|#|$)')
  const results = regex.exec(url)
  if (!results) return null
  if (!results[2]) return ''
  return decodeURIComponent(results[2].replace(/\+/g, ' '))
}

// async function checkUpdateCurrencyInfo () {
//   while (this.engineOn) {
//     try {
//       const url = sprintf('%s/v1/currencyInfo/ETH', INFO_SERVERS[0])
//       const jsonObj = await this.fetchGet(url)
//       const valid = validateObject(jsonObj, CurrencyInfoScheme)
//
//       if (valid) {
//         console.log('Fetched valid currencyInfo')
//         console.log(jsonObj)
//       } else {
//         console.log('Error: Fetched invalid currencyInfo')
//       }
//     } catch (err) {
//       console.log('Error fetching currencyInfo: ' + err)
//     }
//     try {
//       await snooze(BLOCKHEIGHT_POLL_MILLISECONDS)
//     } catch (err) {
//       console.log(err)
//     }
//   }
// }

export const kusdCurrencyPluginFactory: EdgeCurrencyPluginFactory = {
  pluginType: 'currency',
  pluginName: currencyInfo.pluginName,

  async makePlugin (opts: any): Promise<EdgeCurrencyPlugin> {
    io = opts.io

    console.log(`Creating Currency Plugin for kUSD`)
    const kusdPlugin:EdgeCurrencyPlugin = {
      pluginName: 'kusdtestnet',
      currencyInfo,

      createPrivateKey: (walletType: string) => {
        const type = walletType.replace('wallet:', '')

        if (type === 'kusdtestnet') {
          const cryptoObj = {
            randomBytes: randomBuffer
          }
          walletUtils.overrideCrypto(cryptoObj)

          const wallet = walletUtils.generate(false)
          const kusdtestnetKey = wallet.getPrivateKeyString().replace('0x', '')
          return { kusdtestnetKey }
        } else {
          throw new Error('InvalidWalletType')
        }
      },

      derivePublicKey: (walletInfo: EdgeWalletInfo) => {
        const type = walletInfo.type.replace('wallet:', '')
        if (type === 'kusdtestnet') {
          const privKey = hexToBuf(walletInfo.keys.kusdtestnetKey)
          const wallet = walletUtils.fromPrivateKey(privKey)
          const kusdtestnetPublicAddress = wallet.getAddressString()
          // const kusdtestnetKey = '0x389b07b3466eed587d6bdae09a3613611de9add2635432d6cd1521af7bbc3757'
          // const kusdtestnetPublicAddress = '0xd6e579085c82329c89fca7a9f012be59028ed53f'
          return { kusdtestnetPublicAddress }
        } else {
          throw new Error('InvalidWalletType')
        }
      },

      // XXX Deprecated. To be removed once Core supports createPrivateKey and derivePublicKey -paulvp
      createMasterKeys: (walletType: string) => {
        if (walletType === 'kusdtestnet') {
          const cryptoObj = {
            randomBytes: randomBuffer
          }
          walletUtils.overrideCrypto(cryptoObj)

          const wallet = walletUtils.generate(false)
          const kusdtestnetKey = wallet.getPrivateKeyString().replace('0x', '')
          const kusdtestnetPublicAddress = wallet.getAddressString()
          // const kusdtestnetKey = '0x389b07b3466eed587d6bdae09a3613611de9add2635432d6cd1521af7bbc3757'
          // const kusdtestnetPublicAddress = '0xd6e579085c82329c89fca7a9f012be59028ed53f'
          return {kusdtestnetKey, kusdtestnetPublicAddress}
        } else {
          return null
        }
      },

      async makeEngine (walletInfo: EdgeWalletInfo, opts: EdgeCurrencyEngineOptions): Promise<EdgeCurrencyEngine> {
        const engine = new Engine(io, walletInfo, opts)
        try {
          const result =
            await engine.walletLocalFolder
              .folder(DATA_STORE_FOLDER)
              .file(DATA_STORE_FILE)
              .getText(DATA_STORE_FOLDER, 'walletLocalData')

          engine.walletLocalData = new WalletLocalData(result)
          engine.walletLocalData.kusdtestnetAddress = engine.walletInfo.keys.kusdtestnetAddress
        } catch (err) {
          try {
            console.log(err)
            console.log('No walletLocalData setup yet: Failure is ok')
            engine.walletLocalData = new WalletLocalData(null)
            engine.walletLocalData.kusdtestnetAddress = engine.walletInfo.keys.kusdtestnetAddress
            await engine.walletLocalFolder
              .folder(DATA_STORE_FOLDER)
              .file(DATA_STORE_FILE)
              .setText(JSON.stringify(engine.walletLocalData))
          } catch (e) {
            console.log('Error writing to localDataStore. Engine not started:' + err)
          }
        }
        for (const token of engine.walletLocalData.enabledTokens) {
          engine.tokenCheckStatus[token] = 0
        }
        return engine
      },

      parseUri: (uri: string) => {
        const parsedUri = parse(uri)
        let address: string
        let nativeAmount: string | null = null
        let currencyCode: string | null = null

        if (
          typeof parsedUri.scheme !== 'undefined' &&
          parsedUri.scheme !== 'kusdtestnet'
        ) {
          throw new Error('InvalidUriError') // possibly scanning wrong crypto type
        }
        if (typeof parsedUri.host !== 'undefined') {
          address = parsedUri.host
        } else if (typeof parsedUri.path !== 'undefined') {
          address = parsedUri.path
        } else {
          throw new Error('InvalidUriError')
        }
        address = address.replace('/', '') // Remove any slashes
        let [prefix, contractAddress] = address.split('-') // Split the address to get the prefix according to EIP-681
        // If contractAddress is null or undefined it means there is no prefix
        if (!contractAddress) {
          contractAddress = prefix // Set the contractAddress to be the prefix when the prefix is missing.
          prefix = 'pay' // The default prefix according to EIP-681 is "pay"
        }
        address = contractAddress
        const valid: boolean = Util.isValidAddress(address)
        if (!valid) {
          throw new Error('InvalidPublicAddressError')
        }
        // If the address has a "token-" prefix, it means it's an "Add Token" URI and not a payment one.
        if (prefix === 'token' || prefix === 'token_info') {
          const currencyCode = getParameterByName('symbol', uri) || 'SYM'
          if (currencyCode.length < 2 || currencyCode.length > 5) {
            throw new Error('Wrong Token symbol')
          }
          const currencyName = getParameterByName('name', uri) || currencyCode
          const decimalsInput = getParameterByName('decimals', uri) || '18'
          let multiplier = '1000000000000000000'
          try {
            const decimals = parseInt(decimalsInput)
            if (decimals < 0 || decimals > 18) {
              throw new Error('Wrong number of decimals')
            }
            multiplier = '1' + '0'.repeat(decimals)
          } catch (e) {
            throw e
          }

          const type = getParameterByName('type', uri) || 'ERC20'

          const edgeParsedUri: EdgeParsedUri = {
            token: {
              currencyCode,
              contractAddress,
              currencyName,
              multiplier,
              type: type.toUpperCase()
            }
          }
          return edgeParsedUri
        }
        const amountStr = getParameterByName('amount', uri)
        if (amountStr && typeof amountStr === 'string') {
          const denom = getDenomInfo('KUSD')
          if (!denom) {
            throw new Error('InternalErrorInvalidCurrencyCode')
          }
          nativeAmount = bns.mul(amountStr, denom.multiplier)
          nativeAmount = bns.toFixed(nativeAmount, 0, 0)
          currencyCode = 'KUSD'
        }
        const label = getParameterByName('label', uri)
        const message = getParameterByName('message', uri)

        const edgeParsedUri:EdgeParsedUri = {
          publicAddress: address
        }
        if (nativeAmount) {
          edgeParsedUri.nativeAmount = nativeAmount
        }
        if (currencyCode) {
          edgeParsedUri.currencyCode = currencyCode
        }
        if (label || message) {
          edgeParsedUri.metadata = {}
          if (label) {
            edgeParsedUri.metadata.name = label
          }
          if (message) {
            edgeParsedUri.metadata.message = message
          }
        }

        return edgeParsedUri
      },

      encodeUri: (obj: EdgeEncodeUri) => {
        if (!obj.publicAddress) {
          throw new Error('InvalidPublicAddressError')
        }
        const valid: boolean = Util.isValidAddress(obj.publicAddress)
        if (!valid) {
          throw new Error('InvalidPublicAddressError')
        }
        if (!obj.nativeAmount && !obj.label && !obj.message) {
          return obj.publicAddress
        } else {
          let queryString: string = ''

          if (typeof obj.nativeAmount === 'string') {
            let currencyCode: string = 'KUSD'
            const nativeAmount:string = obj.nativeAmount
            if (typeof obj.currencyCode === 'string') {
              currencyCode = obj.currencyCode
            }
            const denom = getDenomInfo(currencyCode)
            if (!denom) {
              throw new Error('InternalErrorInvalidCurrencyCode')
            }
            const amount = bns.div(nativeAmount, denom.multiplier, 18)

            queryString += 'amount=' + amount + '&'
          }
          if (obj.metadata && (obj.metadata.name || obj.metadata.message)) {
            if (typeof obj.metadata.name === 'string') {
              queryString += 'label=' + obj.metadata.name + '&'
            }
            if (typeof obj.metadata.message === 'string') {
              queryString += 'message=' + obj.metadata.message + '&'
            }
          }
          queryString = queryString.substr(0, queryString.length - 1)

          const serializeObj = {
            scheme: 'kusdtestnet',
            path: obj.publicAddress,
            query: queryString
          }
          const url = serialize(serializeObj)
          return url
        }
      }
    }

    if (global.OS && global.OS === 'ios') {
      const metaTokens = []
      for (const metaToken of kusdPlugin.currencyInfo.metaTokens) {
        const currencyCode = metaToken.currencyCode
        if (kusdPlugin.currencyInfo.defaultSettings.otherSettings.iosAllowedTokens[currencyCode] === true) {
          metaTokens.push(metaToken)
        }
      }
      kusdPlugin.currencyInfo.metaTokens = metaTokens
    }

    async function initPlugin (opts: any) {
      // Try to grab currencyInfo from disk. If that fails, use defaults

      // try {
      //   const result =
      //     await this.walletLocalFolder
      //       .folder(DATA_STORE_FOLDER)
      //       .file(DATA_STORE_FILE)
      //       .getText(DATA_STORE_FOLDER, 'walletLocalData')
      //
      //   this.walletLocalData = new WalletLocalData(result)
      //   this.walletLocalData.kusdtestnetAddress = this.walletInfo.keys.kusdtestnetAddress
      // }

      // Spin off network query to get updated currencyInfo and save that to disk for future bootups

      return kusdPlugin
    }
    return initPlugin(opts)
  }
}
